src/vkbd/usb-gadget.c

```diff
@@ -0,0 +1,120 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include "../keys.h"
+#include "usb-gadget.h"
+
+uint16_t mods = 0;
+
+struct vkbd {
+	int fd;
+};
+
+struct hid_report {
+	uint16_t mask;
+	uint16_t code;
+	uint32_t fill;
+};
+
+static int create_virtual_keyboard(void)
+{
+
+	int fd = open("/dev/hidg0", O_WRONLY | O_NONBLOCK);
+	if (fd < 0) {
+		exit(-1);
+	}
+
+	return fd;
+}
+
+struct vkbd *vkbd_init(const char *name)
+{
+	struct vkbd *vkbd = calloc(1, sizeof vkbd);
+	vkbd->fd = create_virtual_keyboard();
+
+	return vkbd;
+}
+
+uint16_t hid_code(uint16_t code)
+{
+	int codes = sizeof(evdev_to_hid)/sizeof(evdev_to_hid[0]);
+	for(int i = 0; i < codes/2; i++) {
+		if(code == evdev_to_hid[2*i]) {
+			return evdev_to_hid[2*i+1];
+		}
+	}
+	return 0;
+}
+
+void send_hid_report (const struct vkbd *vkbd, uint16_t code, uint16_t mods) {
+
+	struct hid_report report;
+
+	report.code = hid_code(code);
+	report.mask = mods;
+	report.fill = 0;
+
+	write(vkbd->fd,&report,sizeof(report));
+
+}
+
+static int get_modifier(const char *name)
+{
+
+	if (!strcmp(name, "leftshift") || !strcmp(name, "rightshift")) {
```

Comment:
There is no need for string comparison here, you should compare the code directly (e.g `code == KEY_RIGHTSHIFT`). A switch statement may be more appropriate in this instance.